core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,289 @@
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Map<String, Schema> inputSchemas = stageConfigurer.getInputSchemas();
+    init();
+    //validate the input schema and get the output schema for it
+    stageConfigurer.setOutputSchema(getOutputSchema(inputSchemas));
+  }
+
+  @Override
+  public void prepareRun(BatchJoinerContext context) throws Exception {
+    if (conf.getNumPartitions() != null) {
+      context.setNumPartitions(conf.getNumPartitions());
+    }
+  }
+
+  @Override
+  public void initialize(BatchJoinerRuntimeContext context) throws Exception {
+    init();
+    inputSchemas = context.getInputSchemas();
+    outputSchema = context.getOutputSchema();
+  }
+
+  @Override
+  public StructuredRecord joinOn(String stageName, StructuredRecord record) throws Exception {
+    List<Schema.Field> fields = new ArrayList<>();
+    Schema schema = record.getSchema();
+
+    List<String> joinKeys = perStageJoinKeys.get(stageName);
+    int i = 1;
+    for (String joinKey : joinKeys) {
+      Schema.Field joinField = Schema.Field.of(String.valueOf(i++), schema.getField(joinKey).getSchema());
+      fields.add(joinField);
+    }
+    Schema keySchema = Schema.recordOf("join.key", fields);
+    StructuredRecord.Builder keyRecordBuilder = StructuredRecord.builder(keySchema);
+    i = 1;
+    for (String joinKey : joinKeys) {
+      keyRecordBuilder.set(String.valueOf(i++), record.get(joinKey));
+    }
+
+    return keyRecordBuilder.build();
+  }
+
+  @Override
+  public JoinConfig getJoinConfig() {
+    return new JoinConfig(requiredInputs);
+  }
+
+  @Override
+  public StructuredRecord merge(StructuredRecord joinKey, Iterable<JoinElement<StructuredRecord>> joinRow) {
+    StructuredRecord.Builder outRecordBuilder;
+    outRecordBuilder = StructuredRecord.builder(outputSchema);
+
+    for (JoinElement<StructuredRecord> joinElement : joinRow) {
+      String stageName = joinElement.getStageName();
+      StructuredRecord record = joinElement.getInputRecord();
+
+      Map<String, String> fieldsToRename = perStageFieldsToRename.row(stageName);
+
+      for (Schema.Field field : record.getSchema().getFields()) {
+        String inputFieldName = field.getName();
+
+        // drop the field if not part of fieldsToRename
+        if (!fieldsToRename.containsKey(inputFieldName)) {
+          continue;
+        }
+
+        String outputFieldName = fieldsToRename.get(inputFieldName);
+        outRecordBuilder.set(outputFieldName, record.get(inputFieldName));
+      }
+    }
+    return outRecordBuilder.build();
+  }
+
+  private void init() {
+    perStageJoinKeys = getJoinKeys();
+    requiredInputs = getRequiredInputs();
+    perStageFieldsToRename = getPerStageFieldsToRename();
+  }
+
+  private Map<String, List<String>> getJoinKeys() {
+    Map<String, List<String>> stageToKey = new HashMap<>();
+    if (!Strings.isNullOrEmpty(conf.getJoinKeys())) {
```

Comment:
I think it would be cleaner if you invert this condition. throw the exception on the next line. that way, the remainder of the function can be un-indented.